### ISSUE TITLE
Q: track remote state using Task

### ISSUE THREAD

**Issue by tmds** [dotnet/runtime#65782]
I have a proxy object that represents state held in a remote application which is accessible via TCP.

It has two methods:
- one to subscribe to an event that gets triggered when the state changes
- one to get the current value of the state.

```cs
class Proxy
{
    void WatchAsync(Action<int> stateChanged);
    Task<int> GetStateAsync();
}
```

I'd like to mirror the state in my local application.
```cs
int state; // local representation of remote state.
// subscribe for changes:
proxy.WatchAsync(v => state = v);
// get the current value:
state = await proxy.GetStateAsync();
```

When the TCP messages are read, they are ordered. That means the the response to `GetStateAsync` is ordered properly among the `WatchAsync` events.

To keep the order in the application, the `WatchAsync` handler needs to be ordered against the `GetStateAsync` continuation.

I don't think it is possible to keep this order when using a `Task` because it can complete on the calling thread when the response is received before the continuation was set.
The chance of this happening is very small, but not impossible.

@stephentoub @davidfowl I wonder if you are familiar with this problem, and maybe have some suggestions on how to deal with it?

**Comment by huoyaoyuan** [comment]
Looking at the api shape, I don't think exposing both change watching and force-querying is a good idea.

Some consideration:
```csharp
Task<Proxy<T>> WatchAsync(); // returns until the first message returned from remote
```

Or you can even wrap the change stream into `IAsyncEnumerable<T>`.

**Comment by tmds** [comment]
> Looking at the api shape, I don't think exposing both change watching and force-querying is a good idea.

The API is D-Bus (https://dbus.freedesktop.org/doc/dbus-specification.html) which is used for IPC on Linux.

The design assumes the user will run its application logic on a single thread, which works well on runtimes that have an eventloop (like nodejs, python, ...).

It may be that the problem is specific for this protocol and not so common in other protocols.

**Comment by stephentoub** [comment]
To make sure I understand, you've got:
```C#
proxy.WatchAsync(v => Work1(v));
var result = await proxy.GetStateAsync();
Work2(result);
```
and you want to make sure that the Work1 is synchronized with Work2 such that a) they don't run concurrently and b) they're ordered with regards to the order of the inputs.  For example, if there were three inputs a/b/c, and both WathAsync was called and then at some point GetStateAsync was called, an expected outcome might be that Work1(a) is invoked, and only once it's completed Work2(b) is invoked, and only once it's completed Work1(c) is invoked?

**Comment by tmds** [comment]
Yes, that is right. Work1 and Work2 sync up the app with the remote state.

A real case is the D-Bus property API which looks like this:

```
method Properties.Get (in STRING interface_name,
                in STRING property_name,
                out VARIANT value);
method Properties.Set (in STRING interface_name,
                in STRING property_name,
                in VARIANT value);
method Properties.GetAll (in STRING interface_name,
                   out ARRAY of DICT_ENTRY<STRING,VARIANT> props);
signal Properties.PropertiesChanged (STRING interface_name,
                              ARRAY of DICT_ENTRY<STRING,VARIANT> changed_properties,
                              ARRAY<STRING> invalidated_properties);
```

**Comment by stephentoub** [comment]
You would need to coordinate the activities using some form of synchronization/scheduling primitive, e.g. the ExclusiveScheduler from a ConcurrentExclusiveSchedulerPair instance.  Basically force all of the activity through a serialized event loop.

**Comment by tmds** [comment]
In order for `GetStateAsync` to use that scheduler, the user of the method should already be on it. So, I cannot hide that this requires a specific scheduler to the user. Is that right?

**Comment by stephentoub** [comment]
> So, I cannot hide that this requires a specific scheduler to the user. Is that right?

If arbitrary code can await your method, that's correct, at least if the method returns a Task.  You could experiment with returning a ValueTask that wraps your own IValueTaskSource implementation, at which point you're entirely responsible for invoking the continuation however, wherever, whenever you see fit.
